src/contracts/pods/EigenPod.sol

```diff
@@ -485,256 +541,108 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
-        
-        /**
-         * Next, calculate the change in number of shares this validator is "backing":
-         * - Anything that needs to go through the withdrawal queue IS backed
-         * - Anything immediately withdrawn IS NOT backed
-         *
-         * This means that this validator is currently backing `amountToQueueGwei` shares.
-         */
-
-        verifiedWithdrawal.sharesDeltaGwei = _calculateSharesDelta({
-            newAmountGwei: amountToQueueGwei,
-            previousAmountGwei: validatorInfo.restakedBalanceGwei
-        });
-
-        /**
-         * Finally, the validator is fully withdrawn. Update their status and place in state:
-         */
-
-        if (validatorInfo.status != VALIDATOR_STATUS.WITHDRAWN) {
-            activeValidatorCount--;
-            validatorInfo.status = VALIDATOR_STATUS.WITHDRAWN;
+        // If the validator was marked stale and this proof is younger than
+        // `TIME_TILL_STALE_BALANCE`, the validator is no longer stale.
+        if (
+            beaconTimestamp + TIME_TILL_STALE_BALANCE >= block.timestamp && 
+            isValidatorStale[proof.pubkeyHash]
+        ) {
+            isValidatorStale[proof.pubkeyHash] = false;
+            noLongerStale = true;
         }
 
-        validatorInfo.restakedBalanceGwei = 0;        
-        _validatorPubkeyHashToInfo[validatorPubkeyHash] = validatorInfo;
-
-        emit FullWithdrawalRedeemed(validatorIndex, withdrawalTimestamp, recipient, withdrawalAmountGwei);
-
-        return verifiedWithdrawal;
+        return (balanceDeltaWei, noLongerStale);
     }
 
-    function _processPartialWithdrawal(
-        uint40 validatorIndex,
-        uint64 withdrawalTimestamp,
-        address recipient,
-        uint64 partialWithdrawalAmountGwei
-    ) internal returns (VerifiedWithdrawal memory) {
-        emit PartialWithdrawalRedeemed(
-            validatorIndex,
-            withdrawalTimestamp,
-            recipient,
-            partialWithdrawalAmountGwei
-        );
-
-        sumOfPartialWithdrawalsClaimedGwei += partialWithdrawalAmountGwei;
-
-        // For partial withdrawals, the withdrawal amount is immediately sent to the pod owner
-        return
-            VerifiedWithdrawal({
-                amountToSendGwei: uint256(partialWithdrawalAmountGwei),
-                sharesDeltaGwei: 0
-            });
+    /**
+     * @dev Finish progress on a checkpoint and store it in state.
+     * @dev If the checkpoint has no proofs remaining, it is finalized:
+     * - a share delta is calculated and sent to the `EigenPodManager`
+     * - the checkpointed `podBalance` is added to `withdrawableRestakedExecutionLayerGwei`
+     * - `lastFinalizedCheckpoint` is updated
+     * - `currentCheckpoint` and `currentCheckpointTimestamp` are deleted
+     */
+    function _updateCheckpoint(Checkpoint memory checkpoint) internal {
+        if (checkpoint.proofsRemaining == 0) {
+            int256 totalShareDelta =
```

Comment:
this isn't a delta?

Reply:
in addition, this needs to be a multiple of gwei (required by EPM), which may be invalidated by selfdestruct. we should probably get rid of that requirement

Reply:
yes it is

Reply:
https://hackmd.io/U36dE9lnQha3tbf7D0GtKw#What-is-a-Checkpoint

Reply:
the question of what to do with any < 1 gwei amount left over is a good question though. i'm not 100% sure what to do with it.